modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTask.java

```diff
@@ -163,94 +205,29 @@ private void performReindex(ReindexJob reindexJob, ReindexTaskStateDoc stateDoc,
         ScrollableHitSource.Checkpoint checkpoint = stateDoc.getCheckpoint();
         ThreadContext threadContext = client.threadPool().getThreadContext();
 
-        // todo: need to store status in state so we can continue from it.
-        if (childTask.isWorker()) { // only unsliced supports restarts.
-            childTask.setCommittedStatus(childTask.getStatus());
-        }
-
-        boolean shouldStoreResult = reindexJob.shouldStoreResult();
         Supplier<ThreadContext.StoredContext> context = threadContext.newRestorableContext(false);
         // TODO: Eventually we only want to retain security context
         try (ThreadContext.StoredContext ignore = stashWithHeaders(threadContext, reindexJob.getHeaders())) {
             reindexer.execute(childTask, reindexRequest, new ContextPreservingActionListener<>(context, new ActionListener<>() {
                 @Override
                 public void onResponse(BulkByScrollResponse response) {
-                    handleDone(shouldStoreResult, taskUpdater, response);
+                    transientStatus = response.getStatus();
+                    taskUpdater.finish(response, null);
                 }
 
                 @Override
-                public void onFailure(Exception ex) {
-                    handleError(shouldStoreResult, taskUpdater, ex);
+                public void onFailure(Exception e) {
+                    taskUpdater.finish(null, wrapException(e));
                 }
-            }), checkpoint, taskUpdater);
+            }), checkpoint, (checkpoint1, status) -> {
```

Comment:
I think we should rename `checkpoint` to `initialCheckpoint` to avoid the name conflict and let `checkpoint1` be just `checkpoint`